modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.server/src/main/scala/edu/gemini/seqexec/web/server/http4s/WebServerLauncher.scala

```diff
@@ -1,38 +1,111 @@
+object WebServerLauncher extends ServerApp with LogInitialization {
   // Initialize the log and exit if it fails
   configLog.run.onException(IO(sys.exit(1))).unsafePerformIO()
 
+  // Initialize logger after the configuration
   val logger = Logger.getLogger(getClass.getName)
 
-  // TODO improve configuration style
-  val devMode = !args.contains("prod")
-
-  def launch(port: Int):Option[Server] = {
-    logger.info(s"Starting web server on port $port")
-    try {
-      Some(BlazeBuilder.bindHttp(port, "0.0.0.0")
-        .withWebSockets(true)
-        .mountService(StaticRoutes.service(devMode), "/")
-        .mountService(SeqexecCommandRoutes.service, "/api/seqexec/commands")
-        .mountService(SeqexecUIApiRoutes.service, "/api")
-        .run)
-    } catch {
-      case e: Throwable =>
-        e.printStackTrace()
-        None
+  /**
+    * Configuration for the web server
+    */
+  case class WebServerConfiguration(host: String, port: Int, devMode: Boolean)
+
+  /**
+    * Configuration for the seqexec engine
+    */
+  case class SeqexecConfiguration(odbHost: String)
+
+  // Attempt to get the file or throw an exception if not possible
+  val configurationFile: File = new File(new File(baseDir, "conf"), "app.conf")
+
+  // Read the config, first attempt the file or default to the classpath file
+  val config: Task[Config] = knobs.loadImmutable(
+    Required(FileResource(configurationFile) or ClassPathResource("app.conf")) :: Nil)
+
+  // configuration specific to the web server
+  val serverConf: Task[WebServerConfiguration] =
+    config.map { cfg =>
+      val host = cfg.require[String]("web-server.host")
+      val port = cfg.require[Int]("web-server.port")
+      val devMode = cfg.require[String]("mode")
+      WebServerConfiguration(host, port, devMode.equalsIgnoreCase("dev"))
+    }
+
+  // Configuration of the ldap clients
+  val ldapConf: Task[LDAPConfig] =
+    config.map { cfg =>
+      val urls = cfg.require[List[String]]("authentication.ldapURLs")
+      LDAPConfig(urls)
+    }
+
+  // Configuration of the authentication service
+  val authConf: Task[AuthenticationConfig] =
+    for {
+      ld <- ldapConf
+      cfg <- config
+    } yield {
+      val devMode = cfg.require[String]("mode")
+      val sessionTimeout = cfg.require[Int]("authentication.sessionLifeHrs")
+      val cookieName = cfg.require[String]("authentication.cookieName")
+      val secretKey = cfg.require[String]("authentication.secretKey")
+      val useSSL = cfg.require[Boolean]("authentication.useSSL")
+      AuthenticationConfig(devMode.equalsIgnoreCase("dev"), sessionTimeout, cookieName, secretKey, useSSL, ld)
+    }
+
+  val executorConf: Task[SeqexecConfiguration] =
+    config.map { cfg =>
+      val host = cfg.require[String]("seqexec-engine.odb")
+      SeqexecConfiguration(host)
     }
+
+  /**
+    * Configures the Seqexec executor
+    */
+  def seqexecExecutor: Kleisli[Task, SeqexecConfiguration, Unit] = Kleisli { conf =>
```

Comment:
Using a `Kleisli` as a ReaderT monad so that the seqexec can access the configuration at any time is perfectly reasonable. The problem is that the seqexec will have its own StateT monad to control the state of execution and I already experienced first hand how painful combining monad transformers of multiple layers is in Scala (the Kleisli and the StateT would have to be stacked).

In any case I'm working now on replacing the StateT monad with a Free Applicative, so we don't need to be concerned with this `Kleisli` now for this PR.

But my point is that we need to at some point to discuss what do we do with multiple effects, do we ditch monad transformers entirely? Do we allow transformers of just one layer only? If we allow them then, which layer takes precedence in this particular case (Kleisli or StateT)?

Sorry for hijacking the thread. We can continue discussing this elsewhere.